File: CloudWS/src/main/java/cloud/logging/OperationLogger.java
package cloud.logging;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.io.File;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class OperationLogger {

    private static final String LOG_FILE = "operations_log.xml";
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss");

    public enum OperationType {
        UPLOAD,
        TELECHARGER,
        SUPPRIMER,
        RESUMER;
    }

    public enum OperationStatus {
        SUCCESS, ERROR
    }
    public static void logOperation(OperationType type, String filename, OperationStatus status, String message, Long fileSize) {
        try {
            Document doc = getOrCreateDocument();
            Element root = doc.getDocumentElement();

            Element operation = doc.createElement("operation");

            String id = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmssSSS")) + "_" + type;
            operation.setAttribute("id", id);

            Element timestamp = doc.createElement("timestamp");
            timestamp.setTextContent(LocalDateTime.now().format(DATE_FORMATTER));
            operation.appendChild(timestamp);

            Element typeElement = doc.createElement("type");
            typeElement.setTextContent(String.valueOf(type));
            operation.appendChild(typeElement);

            Element filenameElement = doc.createElement("filename");
            filenameElement.setTextContent(filename != null ? filename : "");
            operation.appendChild(filenameElement);

            Element statusElement = doc.createElement("status");
            statusElement.setTextContent(status.name());
            operation.appendChild(statusElement);

            Element messageElement = doc.createElement("message");
            messageElement.setTextContent(message != null ? message : "");
            operation.appendChild(messageElement);

            if (fileSize != null && fileSize > 0) {
                Element fileSizeElement = doc.createElement("fileSize");
                fileSizeElement.setTextContent(String.valueOf(fileSize));
                operation.appendChild(fileSizeElement);
            }

            root.appendChild(operation);

            saveDocument(doc);

        } catch (Exception e) {
            System.err.println("Erreur lors de l'enregistrement de l'opération dans le log XML: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static Document getOrCreateDocument() throws ParserConfigurationException, IOException, org.xml.sax.SAXException {
        File logFile = new File(LOG_FILE);
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc;

        if (logFile.exists() && logFile.length() > 0) {
            doc = builder.parse(logFile);
        } else {
            doc = builder.newDocument();
            Element root = doc.createElement("operations");
            root.setAttribute("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance");
            root.setAttribute("xsi:noNamespaceSchemaLocation", "operations_log.xsd");
            doc.appendChild(root);
        }

        return doc;
    }
    private static void saveDocument(Document doc) throws TransformerException {
        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        Transformer transformer = transformerFactory.newTransformer();
        transformer.setOutputProperty(javax.xml.transform.OutputKeys.INDENT, "yes");
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
        transformer.setOutputProperty(javax.xml.transform.OutputKeys.ENCODING, "UTF-8");

        DOMSource source = new DOMSource(doc);
        StreamResult result = new StreamResult(new File(LOG_FILE));
        transformer.transform(source, result);
    }
}

File: AI/operation_logger.py
"""
Module de logging des opérations dans un fichier XML
"""
import xml.etree.ElementTree as ET
from datetime import datetime
from pathlib import Path
from typing import Optional

class OperationLogger:    
    LOG_FILE = "../CloudWS/operations_log.xml"
    DATE_FORMAT = "%Y-%m-%dT%H:%M:%S"
    
    class OperationType:
        SUMMARIZE = "RESUMER"
    
    class OperationStatus:
        SUCCESS = "SUCCESS"
        ERROR = "ERROR"
    
    @staticmethod
    def log_operation(operation_type: str, filename: str, status: str, message: str, file_size: Optional[int] = None):
        root = OperationLogger._get_or_create_root()
        
        operation = ET.SubElement(root, "operation")
        
        operation_id = datetime.now().strftime("%Y%m%d%H%M%S%f")[:-3] + "_" + operation_type
        operation.set("id", operation_id)
        
        timestamp = ET.SubElement(operation, "timestamp")
        timestamp.text = datetime.now().strftime(OperationLogger.DATE_FORMAT)
        
        type_elem = ET.SubElement(operation, "type")
        type_elem.text = operation_type
        
        filename_elem = ET.SubElement(operation, "filename")
        filename_elem.text = filename if filename else ""
        
        status_elem = ET.SubElement(operation, "status")
        status_elem.text = status
        
        message_elem = ET.SubElement(operation, "message")
        message_elem.text = message if message else ""
        
        if file_size is not None and file_size > 0:
            file_size_elem = ET.SubElement(operation, "fileSize")
            file_size_elem.text = str(file_size)
        
        OperationLogger._save_document(root)
    
    @staticmethod
    def _get_or_create_root():
        log_file = Path(OperationLogger.LOG_FILE)
        
        if log_file.exists() and log_file.stat().st_size > 0:
            tree = ET.parse(log_file)
            root = tree.getroot()
            return root
        
        root = ET.Element("operations")
        root.set("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance")
        root.set("xsi:noNamespaceSchemaLocation", "operations_log.xsd")
        return root
    
    @staticmethod
    def _save_document(root):
        log_file = Path(OperationLogger.LOG_FILE)
        ET.indent(root, space="  ")
        tree = ET.ElementTree(root)
        tree.write(log_file, encoding="utf-8", xml_declaration=True)

